type_lens/callable_view.py

```diff
@@ -0,0 +1,80 @@
+from __future__ import annotations
+
+import inspect
+import sys
+import types
+from typing import TYPE_CHECKING, Any, Callable
+
+from typing_extensions import get_type_hints
+
+from type_lens.parameter_view import ParameterView
+from type_lens.type_view import TypeView
+
+__all__ = ("CallableView",)
+
+
+if TYPE_CHECKING:
+    from typing_extensions import Self
+
+
+class CallableView:
+    def __init__(self, fn: Callable, type_hints: dict[str, type]):
+        self.callable = fn
+        self.signature = getattr(fn, "__signature__", None) or inspect.signature(fn)
+
+        return_annotation = type_hints.pop("return", None)
+        self.return_type = TypeView(return_annotation)
+
+        self.parameters = tuple(
+            ParameterView.from_parameter(param, type_hints) for param in self.signature.parameters.values()
+        )
+
+    def __eq__(self, other: object) -> bool:
+        if not isinstance(other, CallableView):
+            return False
+
+        return bool(
+            self.callable == other.callable
+            and self.parameters == other.parameters
+            and self.return_type == other.return_type
+        )
+
+    def __repr__(self) -> str:
+        cls_name = self.__class__.__name__
+
+        return f"{cls_name}({self.callable.__name__})"
+
+    @classmethod
+    def from_callable(
+        cls: type[Self],
+        fn: Callable,
+        *,
+        get_type_hints: Callable = get_type_hints,
+        include_extras: bool = False,
+    ) -> Self:
+        hint_fn = fn
+        if not isinstance(fn, (type, types.FunctionType)):
```

Comment:
With that said, i've renamed it to `CallableView` basically because of this, which only exist to support callable class instances (`test_class_instance`).

I'm not certain this is even case that makes sense to support, but it seemed nice to support for completion's sake.

Reply:
I think it makes sense to support it.

I noticed that an instance method is also not a FunctionType, have added a test and fix in PR to this branch.

Reply:
I xfail'd your failing test examples, since i feel that it doesn't need to block this PR.

I'm not even...certain i think your test cases make sense. they're both examples of things that would fail type checking, without implicit optionals turned on.

For the majority other cases, i believe the `_fix_annotated_optional_type_hints`, that i adapted from cappa i think is covering the very similar cases to what you're referencing.